diff --git a/run/Flan/Modern Weapons Pack/aaguns/Sentry.txt b/run/Flan/Modern Weapons Pack/aaguns/Sentry.txt
index 14e88876..36cffeeb 100644
--- a/run/Flan/Modern Weapons Pack/aaguns/Sentry.txt	
+++ b/run/Flan/Modern Weapons Pack/aaguns/Sentry.txt	
@@ -8,10 +8,10 @@ Model mw.SentryGun
 //Place the texture .png in the skins folder
 Texture SentryGun
 NumBarrels 4
-Barrel 0 40 6 6
-Barrel 1 40 6 6
-Barrel 2 40 6 6
-Barrel 3 40 6 6
+Barrel 0 40 28 16
+Barrel 1 40 28 16
+Barrel 2 40 28 16
+Barrel 3 40 28 16
 FireAlternately False
 //Reload time in ticks, 1/20ths of a second
 ReloadTime 100
diff --git a/src/main/java/com/flansmod/common/FlansMod.java b/src/main/java/com/flansmod/common/FlansMod.java
index 93179288..b96b296a 100644
--- a/src/main/java/com/flansmod/common/FlansMod.java
+++ b/src/main/java/com/flansmod/common/FlansMod.java
@@ -140,7 +140,7 @@ import com.flansmod.common.types.EnumType;
 import com.flansmod.common.types.InfoType;
 import com.flansmod.common.types.TypeFile;
 
-@Mod(modid = FlansMod.MODID, name = "Flan's Mod", version = FlansMod.VERSION, acceptableRemoteVersions = "@ALLOWED_VERSIONS@", guiFactory = "com.flansmod.client.gui.config.ModGuiFactory")
+@Mod(modid = FlansMod.MODID, name = "Flan's Mod KL patch", version = FlansMod.VERSION, acceptableRemoteVersions = "@ALLOWED_VERSIONS@", guiFactory = "com.flansmod.client.gui.config.ModGuiFactory")
 public class FlansMod
 {
 	//Core mod stuff
diff --git a/src/main/java/com/flansmod/common/driveables/EntityDriveable.java b/src/main/java/com/flansmod/common/driveables/EntityDriveable.java
index bb376778..82ec7256 100644
--- a/src/main/java/com/flansmod/common/driveables/EntityDriveable.java
+++ b/src/main/java/com/flansmod/common/driveables/EntityDriveable.java
@@ -79,7 +79,9 @@ public abstract class EntityDriveable extends Entity implements IControllable, I
 	public double serverPosX, serverPosY, serverPosZ;
 	/** Server side rotation, as synced by PacketVehicleControl packets */
 	public double serverYaw, serverPitch, serverRoll;
-	
+
+	private double lastStablePosX = 0, lastStablePosY = 100, lastStablePosZ = 0;
+	public int nanRecoveryAttempt = 0;
 	/** The driveable data which contains the inventory, the engine and the fuel */
 	public DriveableData driveableData;
 	/** The shortName of the driveable type, used to obtain said type */
@@ -192,7 +194,16 @@ public abstract class EntityDriveable extends Entity implements IControllable, I
 			emitterTimers[i] = rand.nextInt(type.emitters.get(i).emitRate);
 		}
 	}
-	
+	public boolean canUpgradeEngine()
+	{
+		return (motionX<1)&&(motionY<1)&&(motionZ<1)&&(fuelling==false);
+	}
+        public PartType replaceEngine(PartType engine)
+        {
+                PartType buf = driveableData.engine;
+                driveableData.engine = engine;
+                return buf;
+        }
 	@SideOnly(Side.CLIENT)
 	@Override
 	public boolean isInRangeToRender3d(double x, double y, double z)
@@ -1225,24 +1236,59 @@ public abstract class EntityDriveable extends Entity implements IControllable, I
 			}
 		}
 	}
-	
-	public void PostUpdate()
+// Applied patch from Flan's mod Orion Rework
+// https://github.com/Lithius0/FlansModOrionRework/commit/b586d325508f3625352fcada6e8770d43758939f
+        public void PostUpdate()
 	{
-		if(Double.isNaN(posX) 
-		|| Double.isNaN(posY)
-		|| Double.isNaN(posZ)
-		|| Float.isNaN(rotationYaw)
-		|| Float.isNaN(rotationPitch)
-		|| !axes.isValid())
+		//Checking validity of position
+		//NaN or infinite values will be picked up and dealt with
+		if(!Double.isFinite(posX) || !Double.isFinite(posY) || !Double.isFinite(posZ))
 		{
-			FlansMod.log.error("Driveable went to NaNsville. Reverting one frame");
-			posX = prevPosX;
-			posY = prevPosY;
-			posZ = prevPosZ;
-			
-			// Just reset the axes
-			axes = new RotatedAxes();
-			prevAxes = new RotatedAxes();
+			if (nanRecoveryAttempt < 5)
+			{
+				FlansMod.log.error("Driveable: " + this.driveableType + " position is invalid! Attempting to revert to previous frame");
+
+				//Resetting velocity. If velocity was NaN, the next position would be NaN, too.
+				motionX = 0;
+				motionY = 0;
+				motionZ = 0;
+				
+				//Also reset the rotation axis, in case that was causing issues.
+				this.axes = new RotatedAxes();
+				
+				if (Double.isFinite(prevPosX) && Double.isFinite(prevPosY) && Double.isFinite(prevPosZ))
+				{
+					posX = prevPosX;
+					posY = prevPosY;
+					posZ = prevPosZ;
+				}
+				else
+				{
+					posX = lastStablePosX;
+					posY = lastStablePosY;
+					posZ = lastStablePosZ;
+					
+					prevPosX = lastStablePosX;
+					prevPosY = lastStablePosY;
+					prevPosZ = lastStablePosZ;
+				}
+				
+				nanRecoveryAttempt++;
+			}
+			else
+			{
+				//TODO: Dump a list of the cargo within the driveable
+				FlansMod.log.error("Too many attempts to recover driveable! Deleting driveable");
+				
+				this.setDead();
+			}
+		} 
+		else 
+		{
+			nanRecoveryAttempt = 0;
+			lastStablePosX = posX;
+			lastStablePosY = posY;
+			lastStablePosZ = posZ;
 		}
 	}
 	
diff --git a/src/main/java/com/flansmod/common/guns/EntityAAGun.java b/src/main/java/com/flansmod/common/guns/EntityAAGun.java
old mode 100644
new mode 100755
index 030bc16e..4a036d98
--- a/src/main/java/com/flansmod/common/guns/EntityAAGun.java
+++ b/src/main/java/com/flansmod/common/guns/EntityAAGun.java
@@ -8,14 +8,19 @@ import net.minecraft.entity.MoverType;
 import net.minecraft.entity.monster.EntityMob;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.entity.EntityLiving;
+import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.SharedMonsterAttributes;
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.potion.PotionEffect;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.EnumHand;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.util.math.Vec3d;
+import net.minecraft.util.text.TextComponentString;
 import net.minecraft.world.World;
 import net.minecraftforge.fml.client.FMLClientHandler;
 import net.minecraftforge.fml.common.network.ByteBufUtils;
@@ -34,7 +39,7 @@ import com.flansmod.common.teams.Team;
 import com.flansmod.common.teams.TeamsManager;
 import com.flansmod.common.vector.Vector3f;
 
-public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
+public class EntityAAGun extends EntityLiving implements IEntityAdditionalSpawnData
 {
 	private int sUpdateTime;
 	private double sPosX;
@@ -52,6 +57,11 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	 * Gun angles
 	 */
 	public float gunYaw, gunPitch;
+	private int gunTiltState;
+	public static final int GUN_TILT_UP=2;
+	public static final int GUN_TILT_GO_UP=1;
+	public static final int GUN_TILT_GO_DOWN=0;
+	public static final int GUN_TILT_DOWN=-1;
 	/**
 	 * Prev gun angles
 	 */
@@ -64,7 +74,7 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	public int currentBarrel; // For cycling through firing each barrel
 	public boolean mouseHeld;
 	public boolean wasShooting;
-	
+
 	//Sentry stuff
 	/**
 	 * Stops the sentry shooting whoever placed it or their teammates
@@ -81,12 +91,30 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	/**
 	 * How often to check for new targets
 	 */
-	public static final float targetAcquireInterval = 10;
+	public static final float targetAcquireInterval = 15;
 	
 	public int ticksSinceUsed = 0;
 	
 	private float yOffset;
+
+	private float targetUpdateTime = 0;
 	
+
+	public static final int AAGUN_TARGET_MOBS_AGGRESSIVE=1;
+	public static final int AAGUN_TARGET_MOBS_FRIENDLY=(1<<1);
+	public static final int AAGUN_TARGET_PLAYERS=(1<<2);
+	public static final int AAGUN_TARGET_VEHICLES=(1<<3);
+
+
+	public static final int AAGUN_MODE_DISABLED = 0,
+            //AAGUN_MODE_MOBS_ALL=(AAGUN_TARGET_MOBS_AGGRESSIVE|AAGUN_TARGET_MOBS_FRIENDLY),
+	    AAGUN_MODE_MOBS_BAD = (AAGUN_TARGET_MOBS_AGGRESSIVE),
+            AAGUN_MODE_PLAYERS = (AAGUN_TARGET_PLAYERS),
+            AAGUN_MODE_DEFENSE = (AAGUN_TARGET_PLAYERS|AAGUN_TARGET_VEHICLES|AAGUN_TARGET_MOBS_AGGRESSIVE),
+            AAGUN_MODE_ALL = (AAGUN_TARGET_MOBS_AGGRESSIVE|AAGUN_TARGET_MOBS_FRIENDLY|AAGUN_TARGET_PLAYERS|AAGUN_TARGET_VEHICLES);
+
+	public int sentryMode;
+
 	public EntityAAGun(World world)
 	{
 		super(world);
@@ -96,6 +124,10 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 		gunYaw = 0;
 		gunPitch = 0;
 		shootDelay = 0;
+		sentryMode = AAGUN_MODE_MOBS_BAD;
+		gunTiltState = GUN_TILT_GO_UP;
+		enablePersistence();
+	        setNoAI(true);	
 	}
 	
 	public EntityAAGun(World world, AAGunType type1, double d, double d1, double d2, EntityPlayer p)
@@ -107,7 +139,15 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 		initType();
 		setPosition(d, d1, d2);
 	}
-	
+
+
+        @Override
+        protected void applyEntityAttributes() {
+                super.applyEntityAttributes();
+                this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(20.0D);
+                this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.0D);
+        }
+
 	@Override
 	public void setPosition(double d, double d1, double d2)
 	{
@@ -144,6 +184,7 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	@Override
 	protected void entityInit()
 	{
+            super.entityInit();
 	}
 	
 	@Override
@@ -177,10 +218,46 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 		return 0D;
 	}
 	
+        @Override
+        public boolean canBeLeashedTo(EntityPlayer player)
+        {
+                return false;
+        }
+
 	public void setMouseHeld(boolean held)
 	{
 		mouseHeld = held;
 	}
+
+	public void changeSentryMode(EntityPlayer entityplayer)
+	{
+		String msg = "";
+		switch(sentryMode)
+		{
+			case(AAGUN_MODE_DISABLED):{
+				sentryMode = AAGUN_MODE_MOBS_BAD;
+				msg = "Sentry is set to attack mobs";
+			}
+			break;
+			case(AAGUN_MODE_MOBS_BAD):{
+				sentryMode = AAGUN_MODE_PLAYERS;
+				msg = "Sentry is set to attack players";
+			}
+			break;
+			case(AAGUN_MODE_PLAYERS):{
+				sentryMode = AAGUN_MODE_DEFENSE;
+				msg = "Sentry is set to defensive mode";
+			}
+			break;
+			default:{
+				sentryMode = AAGUN_MODE_DISABLED;
+				msg = "Sentry is inactive";
+			}
+			break;
+		}
+		target = null;
+		entityplayer.sendMessage(new TextComponentString(msg));
+	}
 	
 	@Override
 	public boolean attackEntityFrom(DamageSource damagesource, float i)
@@ -268,16 +345,28 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 			if(target != null && target.isDead)
 				target = null;
 			//Find a new target if we don't currently have one
-			if(target == null && ticksExisted % targetAcquireInterval == 0)
+			if(target == null && ticksExisted >= targetUpdateTime)
 			{
+				gunTiltState = GUN_TILT_GO_UP;
+
+				targetUpdateTime = ticksExisted + targetAcquireInterval;
 				target = getValidTarget();
 			}
 			if(target != null)
 			{
+				double barrelOffset = (currentBarrel < type.numBarrels) ? type.barrelY[currentBarrel] / 16D : 1.5D;
+
 				double dX = target.posX - posX;
-				double dY = target.posY - (posY + 1.5F);
+				double dY = target.posY - (posY + barrelOffset);
 				double dZ = target.posZ - posZ;
-				
+
+				switch(gunTiltState)
+				{
+					case(GUN_TILT_UP):dY += target.getEyeHeight(); break;
+					case(GUN_TILT_GO_UP):
+					case(GUN_TILT_GO_DOWN):dY += target.getEyeHeight()/2.0D;break;
+				}
+
 				double distanceToTarget = Math.sqrt(dX * dX + dY * dY + dZ * dZ);
 				
 				if(distanceToTarget > type.targetRange)
@@ -286,7 +375,7 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 				{
 					float newYaw = 180F + (float)Math.atan2(dZ, dX) * 180F / 3.14159F;
 					float newPitch = -(float)Math.atan2(dY, Math.sqrt(dX * dX + dZ * dZ)) * 180F / 3.14159F;
-					
+
 					float turnSpeed = 0.25F;
 					
 					gunYaw += (newYaw - gunYaw) * turnSpeed;
@@ -373,7 +462,6 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 			if (target != null || shootPlayer)
 			{
 				EntityPlayerMP player = shootPlayer ? (EntityPlayerMP)getControllingPassenger() : null;
-				
 				for(int j = 0; j < type.numBarrels; j++)
 				{
 					int ammoSlot = j;
@@ -390,6 +478,14 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 						} else
 						{
 							ammo[ammoSlot].setItemDamage(ammo[ammoSlot].getItemDamage() + 1);
+							switch(gunTiltState)
+							{
+								case(GUN_TILT_UP):gunTiltState = GUN_TILT_GO_DOWN; break;
+								case(GUN_TILT_GO_UP):gunTiltState = GUN_TILT_GO_DOWN;break;
+								case(GUN_TILT_GO_DOWN):gunTiltState = GUN_TILT_DOWN;break;
+								case(GUN_TILT_DOWN):
+								default: gunTiltState = GUN_TILT_GO_UP;break;
+							}
 						}
 						shootDelay = type.shootDelay;
 						barrelRecoil[j] = type.recoil;
@@ -397,11 +493,10 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 						Vec3d origin = rotate(type.barrelX[currentBarrel] / 16D - type.barrelZ[currentBarrel] / 16D,
 								type.barrelY[currentBarrel] / 16D,
 								type.barrelX[currentBarrel] / 16D + type.barrelZ[currentBarrel] / 16D).add(posX, posY, posZ);
-						
 						Double radianYaw = Math.toRadians(gunYaw + 90F);
 						Double radianPitch = Math.toRadians(gunPitch);
-						Vector3f shootingDirection = new Vector3f(-Math.sin(radianYaw), Math.cos(radianYaw)*-Math.sin(radianPitch), Math.cos(radianYaw)*Math.cos(radianPitch));
 						
+						Vector3f shootingDirection = new Vector3f(-Math.sin(radianYaw) * Math.cos(radianPitch), -Math.sin(radianPitch), Math.cos(radianYaw)*Math.cos(radianPitch));
 						FireableGun weapon = new FireableGun(type, (float)type.damage, (float)type.accuracy, (float)type.damage, EnumSpreadPattern.circle);
 						FiredShot shot = new FiredShot(weapon, bullet, this, player);
 						//TODO use Vec3d
@@ -423,39 +518,52 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	{
 		return type.targetMobs || type.targetPlayers;
 	}
-	
+
 	public Entity getValidTarget()
 	{
 		if(world.isRemote)
 			return null;
 		if(placer == null && placerName != null)
 			placer = world.getPlayerEntityByName(placerName);
-		for(Object obj : world.getEntitiesWithinAABBExcludingEntity(this, getEntityBoundingBox().expand(type.targetRange, type.targetRange, type.targetRange)))
+		for(Object obj : world.getEntitiesWithinAABBExcludingEntity(this, getEntityBoundingBox().grow(type.targetRange, type.targetRange, type.targetRange)))
 		{
 			Entity candidateEntity = (Entity)obj;
 			
-			if((type.targetMobs && candidateEntity instanceof EntityMob) || (type.targetPlayers && candidateEntity instanceof EntityPlayer))
+			if((sentryMode & AAGUN_TARGET_MOBS_AGGRESSIVE)!=0 && 
+			(candidateEntity instanceof EntityMob ||
+			(candidateEntity instanceof EntityLiving && 
+		        ((EntityLiving)candidateEntity).getAttackTarget() instanceof EntityPlayer && ((EntityLiving)candidateEntity).getAttackTarget() == placer)))
 			{
 				//Check that this entity is actually in range and visible
-				if(candidateEntity.getDistanceSq(this) < type.targetRange * type.targetRange)
-				{
-					if(candidateEntity instanceof EntityPlayer)
-					{
-						if(candidateEntity == placer || candidateEntity.getName().equals(placerName))
-							continue;
-						if(TeamsManager.enabled && TeamsManager.getInstance().currentRound != null && placer != null)
-						{
-							PlayerData placerData = PlayerHandler.getPlayerData(placer, world.isRemote ? Side.CLIENT : Side.SERVER);
-							PlayerData candidateData = PlayerHandler.getPlayerData((EntityPlayer)candidateEntity, world.isRemote ? Side.CLIENT : Side.SERVER);
-							if(candidateData.team == Team.spectators || candidateData.team == null)
-								continue;
-							if(!TeamsManager.getInstance().currentRound.gametype.playerCanAttack((EntityPlayerMP)placer, placerData.team, (EntityPlayerMP)candidateEntity, candidateData.team))
-								continue;
-						}
-					}
+				if(candidateEntity.getDistanceSq(this) < type.targetRange * type.targetRange &&
+				((EntityLivingBase)candidateEntity).canEntityBeSeen(this))
 					return candidateEntity;
+			}
+
+			else if((sentryMode & AAGUN_TARGET_PLAYERS)!=0 && candidateEntity instanceof EntityPlayer)
+			{
+				if(candidateEntity == placer || candidateEntity.getName().equals(placerName))
+					continue;
+				if(TeamsManager.enabled && TeamsManager.getInstance().currentRound != null && placer != null)
+				{
+					PlayerData placerData = PlayerHandler.getPlayerData(placer, world.isRemote ? Side.CLIENT : Side.SERVER);
+					PlayerData candidateData = PlayerHandler.getPlayerData((EntityPlayer)candidateEntity, world.isRemote ? Side.CLIENT : Side.SERVER);
+					if(candidateData.team == Team.spectators || candidateData.team == null)
+						continue;
+					if(!TeamsManager.getInstance().currentRound.gametype.playerCanAttack((EntityPlayerMP)placer, placerData.team, (EntityPlayerMP)candidateEntity, candidateData.team))
+						continue;
 				}
+				if(candidateEntity.getDistanceSq(this) < type.targetRange * type.targetRange &&
+				((EntityLivingBase)candidateEntity).canEntityBeSeen(this))
+					return candidateEntity;
 			}
+			/*else if(sentryMode & AAGUN_TARGET_MOBS_FRIENDLY)
+			{
+				if(candidateEntity instanceof EntityLivingBase && 
+					!(candidateEntity instanceof IMob) &&
+					(candidateEntity.getDistanceSq(this) < type.targetRange * type.targetRange))
+					return candidateEntity;
+			}*/
 		}
 		return null;
 	}
@@ -509,14 +617,21 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 		
 		passenger.setPosition(posX + x2, posY + y, posZ + z2);
 	}
+	// Sentry is immune to potion effects
+	@Override
+	public void addPotionEffect(PotionEffect potioneffectIn)
+	{
+	}
 	
 	@Override
-	protected void writeEntityToNBT(NBTTagCompound nbttagcompound)
+	public void writeEntityToNBT(NBTTagCompound nbttagcompound)
 	{
 		nbttagcompound.setString("Type", type.shortName);
 		nbttagcompound.setInteger("Health", health);
 		nbttagcompound.setFloat("RotationYaw", rotationYaw);
 		nbttagcompound.setFloat("RotationPitch", rotationPitch);
+		nbttagcompound.setInteger("SentryMode", sentryMode);
+		if(isSentry())nbttagcompound.setInteger("TiltState", gunTiltState);
 		for(int i = 0; i < type.numBarrels; i++)
 		{
 			if(ammo[i] != null)
@@ -530,13 +645,15 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	}
 	
 	@Override
-	protected void readEntityFromNBT(NBTTagCompound nbttagcompound)
+	public void readEntityFromNBT(NBTTagCompound nbttagcompound)
 	{
 		type = AAGunType.getAAGun(nbttagcompound.getString("Type"));
 		initType();
 		health = nbttagcompound.getInteger("Health");
 		rotationYaw = nbttagcompound.getFloat("RotationYaw");
 		rotationPitch = nbttagcompound.getFloat("RotationPitch");
+		if(nbttagcompound.hasKey("SentryMode"))sentryMode = nbttagcompound.getInteger("SentryMode");
+		if(nbttagcompound.hasKey("TiltState"))gunTiltState = nbttagcompound.getInteger("TiltState");
 		for(int i = 0; i < type.numBarrels; i++)
 		{
 			ammo[i] = new ItemStack(nbttagcompound.getCompoundTag("Ammo " + i));
@@ -545,7 +662,7 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 	}
 	
 	@Override
-	public boolean processInitialInteract(EntityPlayer entityplayer, EnumHand hand) //interact : change back when Forge updates
+	protected boolean processInteract(EntityPlayer entityplayer, EnumHand hand) //interact : change back when Forge updates
 	{
 		// Player right clicked on gun
 		// Mount gun
@@ -562,6 +679,7 @@ public class EntityAAGun extends Entity implements IEntityAdditionalSpawnData
 			}
 			if(!isSentry())
 				entityplayer.startRiding(this);
+                        else if(entityplayer.getHeldItem(hand) != null && entityplayer.getHeldItem(hand).isEmpty())changeSentryMode(entityplayer);
 			for(int i = 0; i < (type.shareAmmo ? 1 : type.numBarrels); i++)
 			{
 				if(ammo[i] == null  || ammo[i].isEmpty())
diff --git a/src/main/java/com/flansmod/common/guns/EntityDamageSourceFlan.java b/src/main/java/com/flansmod/common/guns/EntityDamageSourceFlan.java
index f15782d7..fb8a8e79 100644
--- a/src/main/java/com/flansmod/common/guns/EntityDamageSourceFlan.java
+++ b/src/main/java/com/flansmod/common/guns/EntityDamageSourceFlan.java
@@ -91,4 +91,15 @@ public class EntityDamageSourceFlan extends EntityDamageSourceIndirect{
 			return new Vec3d(0d, 0d, 0d);
         return super.getDamageLocation();
     }
+
+    @Override
+    @Nullable
+    public Entity getTrueSource() // So that mobs will try to destroy sentry first if it's automatic
+    {
+        if(damageSourceEntity instanceof EntityAAGun)
+        {
+            if(((EntityAAGun)damageSourceEntity).isSentry())return damageSourceEntity;
+        }
+        return super.getTrueSource();
+    }
 }
diff --git a/src/main/java/com/flansmod/common/guns/FiredShot.java b/src/main/java/com/flansmod/common/guns/FiredShot.java
index e432eb19..b85a6c70 100644
--- a/src/main/java/com/flansmod/common/guns/FiredShot.java
+++ b/src/main/java/com/flansmod/common/guns/FiredShot.java
@@ -112,8 +112,9 @@ public class FiredShot {
 	 */
 	public DamageSource getDamageSource(Boolean headshot)
 	{
-		if (player.isPresent()) {
-			return new EntityDamageSourceFlan(weapon.getShortName(), player.get(), player.get(), weapon.getInfoType(), headshot).setProjectile();
+		if (player.isPresent()) { // Make mobs target automatic sentry if present
+                        Entity trueSource = shooter.isPresent() && shooter.get() instanceof EntityAAGun && ((EntityAAGun)shooter.get()).isSentry() ? shooter.get() : player.get();
+			return new EntityDamageSourceFlan(weapon.getShortName(), trueSource, player.get(), weapon.getInfoType(), headshot).setProjectile();
 		}
 		else if(shooter.isPresent())
 		{
diff --git a/src/main/java/com/flansmod/common/guns/boxes/BlockGunBox.java b/src/main/java/com/flansmod/common/guns/boxes/BlockGunBox.java
index 641ed55a..c21e7d23 100644
--- a/src/main/java/com/flansmod/common/guns/boxes/BlockGunBox.java
+++ b/src/main/java/com/flansmod/common/guns/boxes/BlockGunBox.java
@@ -51,7 +51,7 @@ public class BlockGunBox extends Block
 				for(int j = 0; j < inventory.getSizeInventory(); j++)
 				{
 					ItemStack stack = inventory.getStackInSlot(j);
-					if(stack != null && !stack.isEmpty() && stack.getItem() == check.getItem() && stack.getItemDamage() == check.getItemDamage())
+					if(stack != null && !stack.isEmpty() && (stack.getItem() == check.getItem()||entry.hasLogNeeded(stack,check)) && stack.getItemDamage() == check.getItemDamage())
 					{
 						numMatchingStuff += stack.getCount();
 					}
@@ -69,7 +69,7 @@ public class BlockGunBox extends Block
 					for(int j = 0; j < inventory.getSizeInventory(); j++)
 					{
 						ItemStack stack = inventory.getStackInSlot(j);
-						if(amountLeft > 0 && stack != null && !stack.isEmpty() && stack.getItem() == remove.getItem() && stack.getItemDamage() == remove.getItemDamage())
+						if(amountLeft > 0 && stack != null && !stack.isEmpty() && (stack.getItem() == remove.getItem()||entry.hasLogNeeded(stack,remove)) && stack.getItemDamage() == remove.getItemDamage())
 						{
 							amountLeft -= inventory.decrStackSize(j, amountLeft).getCount();
 						}
diff --git a/src/main/java/com/flansmod/common/guns/boxes/GunBoxType.java b/src/main/java/com/flansmod/common/guns/boxes/GunBoxType.java
index 5dd97200..2804af70 100644
--- a/src/main/java/com/flansmod/common/guns/boxes/GunBoxType.java
+++ b/src/main/java/com/flansmod/common/guns/boxes/GunBoxType.java
@@ -4,6 +4,7 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 
+import net.minecraft.init.Blocks;
 import net.minecraft.block.Block;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.Item;
@@ -314,7 +315,10 @@ public class GunBoxType extends BoxType
 			
 			return haveEnoughOf(temporaryInventory, stackNeeded, false);
 		}
-		
+		public boolean hasLogNeeded(ItemStack stackInSlot,ItemStack stackNeeded)
+		{
+			return stackNeeded.getItem()==Item.getItemFromBlock(Blocks.LOG)&&(stackInSlot.getItem()==Item.getItemFromBlock(Blocks.LOG2)||(stackInSlot.getItem()==Item.getItemFromBlock(Blocks.LOG)));
+		}
 		private boolean haveEnoughOf(InventoryPlayer temporaryInventory, ItemStack stackNeeded, boolean takeItems)
 		{
 			//The total amount of items found that match this recipe stack
@@ -325,7 +329,7 @@ public class GunBoxType extends BoxType
 				//Get the stack in each slot
 				ItemStack stackInSlot = temporaryInventory.getStackInSlot(m).copy();
 				//If the stack is what we want
-				if(stackInSlot.getItem() == stackNeeded.getItem() && stackInSlot.getItemDamage() == stackNeeded.getItemDamage())
+				if((stackInSlot.getItem() == stackNeeded.getItem()||hasLogNeeded(stackInSlot,stackNeeded)) && stackInSlot.getItemDamage() == stackNeeded.getItemDamage())
 				{
 					//Work out the amount to take from the stack
 					int amountFound = Math.min(stackInSlot.getCount(), stackNeeded.getCount() - totalAmountFound);
diff --git a/src/main/java/com/flansmod/common/parts/ItemPart.java b/src/main/java/com/flansmod/common/parts/ItemPart.java
index eeefa0dd..48cc9ae5 100644
--- a/src/main/java/com/flansmod/common/parts/ItemPart.java
+++ b/src/main/java/com/flansmod/common/parts/ItemPart.java
@@ -6,10 +6,24 @@ import net.minecraft.client.util.ITooltipFlag;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.EnumActionResult;
 
 import com.flansmod.common.FlansMod;
 import com.flansmod.common.types.IFlanItem;
 import com.flansmod.common.types.InfoType;
+import com.flansmod.common.driveables.EntityDriveable;
+import com.flansmod.common.driveables.EnumDriveablePart;
+import com.flansmod.common.driveables.DriveablePart;
+
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
+import com.flansmod.common.vector.Vector3f;
+
+import net.minecraft.util.EnumHand;
+import net.minecraft.entity.player.EntityPlayer;
+import com.flansmod.client.debug.EntityDebugVector;
 
 public class ItemPart extends Item implements IFlanItem
 {
@@ -30,7 +44,61 @@ public class ItemPart extends Item implements IFlanItem
 		setRegistryName(type.shortName);
 		setCreativeTab(FlansMod.tabFlanParts);
 	}
-	
+	public static ItemPart getFromRegistry(PartType type)
+	{
+		for(ItemPart p:FlansMod.partItems)
+		{
+			if(p.type==type)return p;
+		}
+		return null;
+	}
+	@Override
+	public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer entityplayer, EnumHand hand)
+	{
+		ItemStack itemstack = entityplayer.getHeldItem(hand);
+		if(type.category == EnumPartCategory.ENGINE)
+		{
+			//Raytracing
+			float cosYaw = MathHelper.cos(-entityplayer.rotationYaw * 0.01745329F);
+			float sinYaw = MathHelper.sin(-entityplayer.rotationYaw * 0.01745329F);
+			float cosPitch = -MathHelper.cos(entityplayer.rotationPitch * 0.01745329F);
+			float sinPitch = MathHelper.sin(entityplayer.rotationPitch * 0.01745329F);
+			double length = 5D;
+			Vec3d posVec = new Vec3d(entityplayer.posX, entityplayer.posY + 1.62D - entityplayer.getYOffset(), entityplayer.posZ);
+			Vec3d lookVec = posVec.add(sinYaw * cosPitch * length, sinPitch * length, cosYaw * cosPitch * length);
+			
+			if(world.isRemote && FlansMod.DEBUG)
+			{
+				world.spawnEntity(new EntityDebugVector(world, new Vector3f(posVec), new Vector3f(posVec.subtract(lookVec)), 100));
+			}
+			
+			//Iterate over all EntityDriveables
+			for(int i = 0; i < world.loadedEntityList.size(); i++)
+			{
+				Object obj = world.loadedEntityList.get(i);
+				if(obj instanceof EntityDriveable)
+				{
+					EntityDriveable driveable = (EntityDriveable)obj;
+					//Raytrace
+					DriveablePart part = driveable.raytraceParts(new Vector3f(posVec), Vector3f.sub(new Vector3f(lookVec), new Vector3f(posVec), null));
+					//If we hit something that is an engine
+					if(part != null&&part.type==EnumDriveablePart.core&&driveable.canUpgradeEngine())
+					{
+						itemstack.setCount(itemstack.getCount() - 1);
+						PartType oldengine = driveable.replaceEngine(type);
+						ItemPart oldpart = getFromRegistry(oldengine); // drop old engine
+						if(oldpart!=null)
+						{
+							ItemStack oldstack = new ItemStack(oldpart,1);
+							entityplayer.dropItem(oldstack,false);
+							return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+						}
+					}
+				}
+			}
+		}
+		return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+	}
 	@Override
 	public void addInformation(ItemStack stack, World world, List<String> lines, ITooltipFlag b)
 	{
diff --git a/src/main/java/com/flansmod/common/types/InfoType.java b/src/main/java/com/flansmod/common/types/InfoType.java
index d482b984..77932f8b 100644
--- a/src/main/java/com/flansmod/common/types/InfoType.java
+++ b/src/main/java/com/flansmod/common/types/InfoType.java
@@ -35,6 +35,7 @@ import net.minecraftforge.registries.IForgeRegistry;
 
 import com.flansmod.common.FlansMod;
 import com.flansmod.common.driveables.DriveableType;
+import net.minecraftforge.fml.common.Loader;
 
 public class InfoType
 {
@@ -741,8 +742,11 @@ public class InfoType
 		AddOreDictEntry("ingotBronze", Ingredient.fromItem(Items.GOLD_INGOT));
 		AddOreDictEntry("blockBronze", Ingredient.fromItems(Item.getItemFromBlock(Blocks.GOLD_BLOCK)));
 
+                AddModEntry("highexgunpowder", "flansmod:gunpowderblock", Ingredient.fromItems(Items.BLAZE_POWDER));
+
 		// IE lookups
-		AddModEntry("treatedPlanks", "immersiveengineering:treated_wood",  Ingredient.fromItems(Item.getItemFromBlock(Blocks.PLANKS)));
+		if(Loader.isModLoaded("immersiveengineering"))AddModEntry("treatedPlanks", "immersiveengineering:treated_wood",  Ingredient.fromItems(Item.getItemFromBlock(Blocks.PLANKS)));
+                else AddModEntry("treatedPlanks","minecraft:chest",Ingredient.fromItems(Item.getItemFromBlock(Blocks.CHEST)));
 	}
 	
 	private static void AddModEntry(String name, String resLoc, Ingredient fallback)
